### PR TITLE
Feat: Add BackboneLambdaFinetunningCallback

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR adds a Fine-tuning Callback + a new callback `on_before_accelerator_backend_setup` + update fine-tuning example to use the new Callback. The example was broken and didn't converge.

Screenshot 2021-01-07 at 15 12 08

Fixes # (issue) <- this [links related issue to this PR](https://docs.github.com/en/free-pro-team@latest/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified
 - [ ] **Check that target branch and milestone match!**

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/callbacks/__init__.py [hunk 1]
```diff
@@ -13,18 +13,19 @@
 # limitations under the License.
 from pytorch_lightning.callbacks.base import Callback
 from pytorch_lightning.callbacks.early_stopping import EarlyStopping
+from pytorch_lightning.callbacks.finetunning import BackboneLambdaFinetunningCallback
```
Comment (justusschock) [inline pytorch_lightning/callbacks/__init__.py:16]:
```suggestion
from pytorch_lightning.callbacks.finetunning import BackboneLambdaFinetunningCallback, BaseFineTuningCallback
```

File: pytorch_lightning/callbacks/__init__.py [hunk 2]
```diff
@@ -13,18 +13,19 @@
...
 from pytorch_lightning.callbacks.model_checkpoint import ModelCheckpoint
 from pytorch_lightning.callbacks.progress import ProgressBar, ProgressBarBase
 
-
 __all__ = [
     'Callback',
     'EarlyStopping',
     'GPUStatsMonitor',
     'GradientAccumulationScheduler',
+    'BackboneLambdaFinetunningCallback',
```
Comment (justusschock) [inline pytorch_lightning/callbacks/__init__.py:28]:
```suggestion
    'BackboneLambdaFinetunningCallback',
    'BaseFineTuningCallback',
```

**Review by justusschock — APPROVED** [review]

**Comment by justusschock** [comment]
Looks good to me, I just would also expose the `BaseFineTuningCallback` as we also expose `ProgressbarBase`

File: pl_examples/domain_templates/computer_vision_fine_tuning.py [hunk 3]
```diff
@@ -58,107 +56,45 @@
...
+    def freeze_before_training(self, pl_module: pl.LightningModule):
+        freeze(module=pl_module.feature_extractor, train_bn=self.train_bn)
 
-    Args:
-        module: A given module
-        train_bn: If True, leave the BatchNorm layers in training mode
+    def finetunning_function(self, pl_module: pl.LightningModule, epoch: int, optimizer: Optimizer, opt_idx: int):
+        if epoch == self.milestones[0]:
+            unfreeze_and_add_param_group(
+                module=pl_module.feature_extractor[-5:],
```
Comment (SeanNaren) [inline pl_examples/domain_templates/computer_vision_fine_tuning.py:80]: Should we add some info of what is actually be unfriozen here?

**Review by SeanNaren — APPROVED** [review]
Aside from small spelling nits, everything looks good to me! Also like the new hook, was thinking if we could simplify the name to `on_before_accelerator_setup` but either way, more hooks the better!

File: tests/callbacks/test_finetunning_callback.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+def test_finetunning_callback(tmpdir):
+    """Test finetunning callbacks works as expected"""
+
+    seed_everything(42)
+
+    class FinetunningBoringModel(BoringModel):
+        def __init__(self):
+            super().__init__()
+            self.backbone = nn.Sequential(nn.Linear(32, 32, bias=False), nn.BatchNorm1d(32), nn.ReLU())
```
Comment (Borda) [inline tests/callbacks/test_finetunning_callback.py:30]: does it work without a backbone, with pure `BoringModel`?

Comment (tchaton) [reply tests/callbacks/test_finetunning_callback.py:30]: No, it expects a module `backbone`.

File: pytorch_lightning/callbacks/finetuning.py [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+    def unfreeze_and_add_param_group(
+        module: Module,
+        optimizer: Optimizer,
+        lr: Optional[float] = None,
+        train_bn: bool = True
+    ):
+        """Unfreezes a module and adds its parameters to an optimizer."""
+        _make_trainable(module)
+        params_lr = optimizer.param_groups[0]['lr'] if lr is None else float(lr)
+        denom_lr = 10. if lr is None else 1.
```
Comment (carmocca) [inline pytorch_lightning/callbacks/finetuning.py:121]: why 10?

File: pytorch_lightning/callbacks/finetuning.py [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+
+def multiplicative(epoch):
+    return 2
+
+
+class BaseFinetuningCallback(Callback):
+
+    r"""
+    BaseFinetuningCallback.
+    Overrides the finetunning_function and add your own logic there.
```
Comment (carmocca) [inline pytorch_lightning/callbacks/finetuning.py:95]:
```suggestion
    Override the finetunning_function and add your own logic there.
```

Can you also mention that it is done `on_epoch_start`?

File: pytorch_lightning/callbacks/finetuning.py [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+
+from pytorch_lightning import _logger as log
+from pytorch_lightning.callbacks.base import Callback
+from pytorch_lightning.core.lightning import LightningModule
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+
+BN_TYPES = (torch.nn.BatchNorm1d, torch.nn.BatchNorm2d, torch.nn.BatchNorm3d)
+
+
+def _make_trainable(module: Module) -> None:
```
Comment (carmocca) [inline pytorch_lightning/callbacks/finetuning.py:35]: Why don't you put these functions inside of the Callback? Users might want to override them

File: tests/trainer/logging/test_logger_connector.py [hunk 8]
```diff
@@ -330,7 +332,7 @@ def test_call_back_validator(tmpdir):
     ]
 
     assert (
-        funcs_name == callbacks_func
+        sorted(funcs_name) == sorted(callbacks_func)
```
Comment (carmocca) [inline tests/trainer/logging/test_logger_connector.py:335]: I believe funcs_name is already sorted

```suggestion
        funcs_name == sorted(callbacks_func)
```

File: pytorch_lightning/callbacks/base.py [hunk 9]
```diff
@@ -27,6 +27,10 @@ class Callback(abc.ABC):
     Subclass this class and override any of the relevant hooks
     """
 
+    def on_before_accelerator_backend_setup(self, trainer, pl_module):
```
Comment (carmocca) [inline pytorch_lightning/callbacks/base.py:30]: do we need to add a new callback hook? can't it be done with the existing ones?

File: pytorch_lightning/callbacks/finetuning.py (hunk 5 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/finetuning.py:121]: It was a default. Not special reason. Let me make it an argument.

File: pytorch_lightning/callbacks/base.py (hunk 9 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/base.py:30]: Great question ! 

We have `on_init_start, on_init_end` in __init__ function. We don't have access to the model.

Then we have on_fit_start in fit function, but `configure_optimizers` had already being called during accelerator_backend.setup().

Therefore, I needed to introduce a new callback, so I could freeze the model before `configure_optimizers` call and then filter parameters if they don't require gradients with
`optimizer = optim.Adam(filter(lambda p: p.requires_grad, self.parameters()), lr=self.lr)`

Hope it answer your question.

File: pytorch_lightning/callbacks/finetuning.py (hunk 5 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/finetuning.py]: Done !

File: pytorch_lightning/callbacks/finetuning.py (hunk 7 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/finetuning.py]: Done !

File: pytorch_lightning/callbacks/base.py (hunk 9 above)
Comment (carmocca) [reply pytorch_lightning/callbacks/base.py:30]: In that case, I think `on_before_configure_optimizers` would be a better name. Easier for users to understand.

Can you also mention the added hook in the CHANGELOG?

**Review by williamFalcon — APPROVED** [review]
